sky/server/server.py

```diff
@@ -267,6 +273,112 @@ async def dispatch(self, request: fastapi.Request, call_next):
+    """Middleware to handle Bearer Token Auth (Service Accounts)."""
+
+    async def dispatch(self, request: fastapi.Request, call_next):
+        # Only process requests with the a path prefix /sa
+        if not request.url.path.startswith('/sa/'):
+            return await call_next(request)
+
+        # Check if service account tokens are enabled
+        sa_enabled = os.environ.get(constants.ENV_VAR_ENABLE_SERVICE_ACCOUNTS,
+                                    'false').lower()
+        if sa_enabled != 'true':
+            logger.warning('Service account authentication disabled')
+            return fastapi.responses.JSONResponse(
+                status_code=401,
+                content={'detail': 'Service account authentication disabled'})
+
+        auth_header = request.headers.get('authorization')
+        if not auth_header or not auth_header.lower().startswith('bearer '):
+            return fastapi.responses.JSONResponse(
+                status_code=401,
+                content={
+                    'detail': 'Bearer token required for service account access'
+                })
+
+        # Extract and validate service account token
+        sa_token = auth_header.split(' ', 1)[1]
+        if not sa_token.startswith('sky_'):
+            return fastapi.responses.JSONResponse(
+                status_code=401,
+                content={'detail': 'Invalid service account token format'})
+
+        try:
+            # Import here to avoid circular imports
+            # pylint: disable=import-outside-toplevel
+            from sky.users.token_service import token_service
+
+            # Verify and decode JWT token
+            payload = token_service.verify_token(sa_token)
+
+            if payload is None:
+                logger.warning(
+                    'BearerTokenMiddleware: Token verification failed')
+                return fastapi.responses.JSONResponse(
+                    status_code=401,
+                    content={
+                        'detail': 'Invalid or expired service account token'
+                    })
+
+            # Extract user information from JWT payload
+            user_id = payload.get('sub')  # Subject = service account user ID
+            user_name = payload.get('name')
+            token_id = payload.get('token_id')
+
+            if not user_id or not token_id:
+                logger.warning(
+                    'Invalid token payload: missing user_id or token_id')
+                return fastapi.responses.JSONResponse(
+                    status_code=401,
+                    content={'detail': 'Invalid token payload'})
+
+            # Verify user still exists in database
+            user_info = global_user_state.get_user(user_id)
+            if user_info is None:
+                logger.warning(f'Token user {user_id} no longer exists')
+                return fastapi.responses.JSONResponse(
+                    status_code=401,
```

Comment:
Do we need to check the service_account_token table as well?

Reply:
currently, the service account token and the service account has 1-1 mapping, i.e. they will be created and removed at the same time, so probably no need to check separately